DebugMode using Particles & Comments

lzardy opened the pull request:
Adds the !debugmode command which has the option to draw the player's server-side collision radius and waypoints. This can be expanded to other units such as projectiles and minions.
Streamlined the process of adding, removing, and networking particles.
Added more variables to particles related to networking.
Cleaned the GameObject.Move function and added some functions related to movement.
Cleaned the NotifyFXCreateGroup function and used new Particle variables.
Added comments to GameObject and Particle.
Shifted location of adding players in Game so that PacketNotifier can never cause an exception.

Resolves #309

lzardy committed:
DebugMode using Particles & Comments

Adds the !debugmode command which has the option to draw the player's server-side collision radius and waypoints. This can be expanded to other units such as projectiles and minions.
Streamlined the process of adding, removing, and networking particles.
Added more variables to particles related to networking.
Cleaned the GameObject.Move function and added some functions related to movement.
Cleaned the NotifyFXCreateGroup function and used new Particle variables.
Added comments to GameObject and Particle.
Shifted location of adding players in Game so that PacketNotifier never causes an exception.

danil179 commented:
I'm merging PR #971 as #966 seems to not occur anymore (they probably have fixed it). That will probably create some merge conflicts. So after I'll end merging you will need to simply rebase (it shouldn't take long to rebase it and it shouldn't be complicated, be sure to keep always another backup branch in case something goes wrong).

lzardy commented:
Rebase successful.

lzardy committed:
Debug Champions & Fix Projectiles

Adds the option to debug all champions via "!debugmode champions".
Added GetAllChampions() function in ObjectManager for the added debug mode.
Fixes projectiles not properly utilizing the Move() function, however this forced Move() to be dependent on Target again.

lzardy commented:
Closing to test travis re-build on re-PR.

danil179 commented on GameServerLib/Chatbox/Commands/DebugModeCommand.cs:
```diff
@@ -0,0 +1,307 @@
+            // Creates a blue flashing highlight around your unit
+            _game.PacketNotifier.NotifyCreateUnitHighlight(userId, _userChampion);
+
+            for (var i = 0.0f; i < 86400.00f && _debugMode == 2; i += 0.1f) //debug for 24 hours, drawing every 100 milliseconds
```
This is not how it should be done.

danil179 commented on GameServerLib/GameObjects/GameObject.cs:
```diff
@@ -91,33 +94,43 @@ public virtual void OnCollision(IGameObject collider)
+        public void Move(float diff, bool useTarget = false)
         {
+            // no waypoints remained - clear the Waypoints
+            if (WaypointIndex >= Waypoints.Count)
```
Use your function `IsPathEnded`.

danil179 commented on GameServerLib/GameObjects/GameObject.cs:
```diff
@@ -91,33 +94,43 @@ public virtual void OnCollision(IGameObject collider)
+            // no waypoints remained - clear the Waypoints
+            if (WaypointIndex >= Waypoints.Count)
+            {
+                Waypoints.RemoveAll(v => v != Waypoints[0]);
```
Calling RemoveAll every time is probably not doing good to the CPU. Maybe add to the condition ` && Waypoints.Count > 1`.

danil179 reviewed (approved):
Approved. Some minor comments.
